Feature login attempts

phgage opened the pull request:
I think if the new Database class is going to pull a copy from the SQL database (edit: Mat has corrected me, this just pulls the data rather than copy) then I thought probably I could go ahead and write this without writing in the SQL for now?

I have edited it so that a user has only 3 login attempts, it will stop login attempts until the user reconnects. If the same username is used 3 times, it will change a Customer variable accountLocked to true.

phgage commented on server/NewBankClientHandler.java:
```diff
@@ -105,4 +119,47 @@ public void sendOutput(String s) {
+
+		//if number of times username has been attempted == maxAttempts then returns username
+		for (Map.Entry<String, Integer> set : countMap.entrySet()){
+			if (set.getValue() == maxAttempts){
```
have just realised this should be >=maxAttempts. Will change it after review.

MattKomorcec commented on server/Customer.java:
```diff
@@ -59,4 +61,17 @@ public String getUsername() {
+		this.accountLocked = true;
+	}
+
+	public void setUnlockAccount(){
```
`setAccountLocked()` and `setUnlockAccount()` could have been just one function, something like:

```java
public void setAccountLocked(boolean value) {
  accountLocked = value;
}
```

I'm not sure which way I like more, just dropping ideas :)

MattKomorcec commented on server/NewBank.java:
```diff
@@ -87,4 +87,8 @@ public String getID(Customer c) {
 		return "";
 	}
+
+	public Customer getCustomer(String s){
```
What is `s`? You should give your parameters meaningful names, in 99% of cases. 1% is for example when you have a for loop, and name a variable `i`

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -23,46 +29,64 @@ public void run() {
+							out.println("Enter Username:");
+							String userName = in.readLine();
+
+							/*
```
This comment should go above the `isAccountLocked` function, not here

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -23,46 +29,64 @@ public void run() {
+								showCustomerMenu(customer);
+								while (true) {
+									String request = in.readLine();
+									System.out.println("Request from " + customer.getUsername());
```
Here you use `System.out.println`

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -23,46 +29,64 @@ public void run() {
+									System.out.println("Request from " + customer.getUsername());
+									// Added pass newBankClientHandler Object into processRequest, so it can call methods
+									String response = bank.processRequest(customer, request, this);
+									out.println(response);
```
Here you use just the `out.println`, and it goes like this through the function. You should decide which version you want to use, and then be consistent with it, as it can cause confusion

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -105,4 +129,51 @@ public void sendOutput(String s) {
+	}
+
+	//counts the frequency of usernames for which logins have been attempted and checks against maximum attempts set
+	private String userLogInFrequency(ArrayList<String> failedLogInUsers, int maxAttempts){
```
This function doesn't really need `failedLogInUsers` as a parameter, because it's already part of the class, it can just access it

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -23,46 +29,64 @@ public void run() {
+								out.println("Loading...");
+								out.println("Log In Failed.\n");
+								//increases failed log in count by one and adds the string of username to failed login list.
+								failedLogInCount += 1;
```
The way you implemented this wouldn't really work as desired from my understanding. `failedLogInCount` tracks the total number of failed logins, not failed logins per user, so for example, I first try to login as Matija, fail 2 times, then I try to login as Stephen, fail once, and that's it, I can't login anymore, I have to disconnect and try again.

You could implement something like `countMap` in `userLogInFrequency`, and store usernames and number of failed logins, and then decide depending on that.

phgage replied:
I think my reasoning was to have two different mechanisms, one for total login attempts and one for username locking.
This is to prevent a user from attempting as many login attempts into different accounts as they liked with say '1234' and hoping to crack one. So we could set the failedLogInCount as 5, but still have an account lock after 3 attempts. 
It still does lock the user if you key in the same name 3 times.

I have made some changes:
1. I have removed the failedLogInCount as failedLogInUsers.size() does the same thing. I have set the total login attempts to 5 now so you can keep trying to login after locking one account.
2. userLogInFrequency() now returns the countMap as I wanted to keep the method to doing one thing
3. added lockUsers which takes that countMap, and for every username, it checks the frequency and if the account is not yet locked, locks it and displays a message.

MattKomorcec reviewed (commented):
Left some comments, please let me know if you need any help

MattKomorcec commented:
It seems like there are some merge conflicts, you should update your local main, and then merge it into this feature branch. If you need help resolving merge conflicts, let me know

sg2550 commented:
Hi guys.

Just had a look at the code and Mat's comments. Mat thanks for your comments. I totally agree with all of the, Great spot regarding the "failedLogInCount". 

Stephen are you happy with the above comments? 

Many thanks - Spyros

sg2550 reviewed (changes requested):
Hi guys.

Just had a look at the code and Mat's comments. Mat thanks for your comments. I totally agree with all of the, Great spot regarding the "failedLogInCount".

Stephen are you happy with those comments?

Many thanks - Spyros

phgage commented:
Thanks all for comments, have made some changes:

1. I have removed the failedLogInCount as failedLogInUsers.size() does the same thing. I have set the total login attempts to 5 now so you can keep trying to login after locking one account.
2. userLogInFrequency() now returns the countMap as I wanted to keep the method to doing one thing
3. added lockUsers which takes that countMap, and for every username, it checks the frequency and if the account is not yet locked, locks it and displays a message.

Also addressed all the other comments as well. 

Hopefully it is now making a bit more sense, let me know what you think. Thanks!

MattKomorcec commented on server/Customer.java:
```diff
@@ -4,6 +4,8 @@
 
 public class Customer {
 
+	private boolean accountLocked = false;
```
Assigning `false` here is not needed, as the default value for a `boolean` is already `false`. All types have their defaults, for example, default values for objects (created from classes) is `null`. 

If you had an object of `Customer` type, this is the same:

```java
Customer customer = null;
Customer customer;
```

MattKomorcec commented on server/Customer.java:
```diff
@@ -62,4 +64,13 @@ public String getUsername() {
 	}
+
+	public void setAccountLocked(boolean value){
+		this.accountLocked = value;
```
Just a nitpick, but you said you wanted me to give you comments :)

Similar to how you used `System.out` and `out`, here you use `this.accountLocked` and `accountLocked`, decide for one and stay consistent

MattKomorcec commented on server/NewBankClientHandler.java:
```diff
@@ -118,4 +147,41 @@ public String printCustomerMenu () {
+
+		//iterates through the map of login frequencies
+		//if user login number>maxAttempts and the account is not already locked, it locks it and prints message
+		for (Map.Entry<String, Integer> set : countMap.entrySet()){
```
This `for` loop got a bit messy and hard to read. Since you use some things multiple times, you could have them as variables. Also, try to get into a habit of re-formatting your code regularly :)

Suggestion:

```java
for (Map.Entry<String, Integer> set : countMap.entrySet()) {
        Integer userAttempts = set.getValue();
        String customer = set.getKey();
        List<Customer> customers = bank.getBank().getCustomers();
        
	if (userAttempts >= maxAttempts) {
		if (customers.containsKey(customer) && !isAccountLocked(customer)) {
			bank.getCustomer(customer).setAccountLocked(true);
			out.println("The account: " + customer + " has been locked.\n");
		}
	}
}
```

MattKomorcec reviewed (commented):
A few more comments to address

MattKomorcec commented:
@phgage now ti makes more sense, after you explained, thanks! It's a bit different to what is usually being done with locking accounts, so I thought you wanted it to work the same, but your way works perfectly fine as well :)

phgage commented:
Thanks Matt and Spyros, agree and have added all of Matt's latest suggestions so hopefully ready to merge

portiamcd1 commented:
Hi, 
I just had a look at the most recent commit, great work @phgage  It is looking really good!! At this time, I cannot see any issues at all :)
